stock_valuation_fifo_lot/models/stock_move_line.py

```diff
@@ -0,0 +1,89 @@
+# Copyright 2024 Quartile (https://www.quartile.co)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)
+
+from odoo import api, fields, models
+
+
+class StockMoveLine(models.Model):
+    _inherit = "stock.move.line"
+
+    qty_consumed = fields.Float(
+        help="Quantity that has gone out of the inventory for valued incoming moves "
+        "for FIFO products with a lot/serial.",
+    )
+    qty_remaining = fields.Float(
+        compute="_compute_remaining_value",
+        store=True,
+        help="Remaining quantity for valued incoming moves for FIFO products with a "
+        "lot/serial.",
+    )
+    company_currency_id = fields.Many2one(related="company_id.currency_id")
+    value_consumed = fields.Monetary(
+        currency_field="company_currency_id",
+        help="The value of the inventory that has been consumed for FIFO products with "
+        "a lot/serial.",
+    )
+    value_remaining = fields.Monetary(
+        compute="_compute_remaining_value",
+        store=True,
+        currency_field="company_currency_id",
+    )
+    force_fifo_lot_id = fields.Many2one(
+        "stock.lot",
+        "Force FIFO Lot/Serial",
+        help="Specify a lot/serial to be consumed (in FIFO costing terms) for the "
+        "outgoing move line, in case the selected lot has already gone out of stock "
+        "(in FIFO costing terms).",
+    )
+
+    @api.depends(
+        "qty_done", "qty_consumed", "move_id.stock_valuation_layer_ids.remaining_value"
+    )
+    def _compute_remaining_value(self):
+        for rec in self:
+            if (
+                rec.product_id.with_company(rec.company_id.id).cost_method != "fifo"
+                or not rec.lot_id
+            ):
+                continue
+            if rec.location_usage in (
+                "internal",
+                "transit",
+            ) or rec.location_dest_usage not in ("internal", "transit"):
+                layers = rec.move_id.stock_valuation_layer_ids
+                remaining_qty_layers = layers.filtered(lambda l: l.remaining_qty > 0)
+                if not remaining_qty_layers:
+                    rec.qty_remaining = 0
+                    rec.value_remaining = 0
+                    continue
+                rec.qty_remaining = rec.product_uom_id._compute_quantity(
+                    sum(remaining_qty_layers.mapped("remaining_qty")),
+                    rec.product_id.uom_id,
+                )
+                rec.value_remaining = (
+                    sum(remaining_qty_layers.mapped("remaining_value"))
+                    * sum(remaining_qty_layers.mapped("remaining_qty"))
+                    / rec.qty_remaining
+                )
+                continue
```

Comment:
We should update the README as well to explain the design.

```suggestion
                # Specifically for the case where qty_done of the outgoing stock move line with done state is reduced,
                # which creates a positive stock valuation layer for the outgoing move.
                layers = rec.move_id.stock_valuation_layer_ids.filtered(lambda l: l.remaining_qty > 0)
                if not layers:
                    continue
                rec.qty_remaining = rec.product_id.uom_id._compute_quantity(
                    sum(layers.mapped("remaining_qty")), rec.product_uom_id
                )
                rec.value_remaining = (
                    sum(layers.mapped("remaining_value"))
                    * sum(layers.mapped("remaining_qty"))
                    / rec.qty_remaining
                )
                continue
```

Reply:
We still need this code for the case of there is no remaining_qty in SVL.
Eg, 
- Receive 10
- Deliver 10
- Reduce deliver qty_done to 8
At the moment, the remaining qty of outgoing move line is 2.
Then,
- Deliver 2 
In this case, we need to clear existing remaining_qty.
```
if not remaining_qty_layers:
   rec.qty_remaining = 0
   rec.value_remaining = 0
   continue
```